gobblin-metrics/src/main/java/gobblin/metrics/RootMetricContext.java

```diff
@@ -163,4 +166,22 @@ protected void addMetricContext(MetricContext context) {
     this.sendNotification(new NewMetricContextNotification(context, context.getInnerMetricContext()));
   }
 
+  /**
+   * Add a shutwon hook that first closes the {@link RootMetricContext} and then invokes {@link #stopReporting()}. This
+   * ensures any resources obtained by the {@link RootMetricContext} are released, and that all reporting started on the
+   * {@link RootMetricContext} stops properly.
+   */
+  private void addShutdownHook() {
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        try {
+          close();
```

Comment:
Closing before calling a method is a weird pattern. Does it need to be done in this order?